File: java/3SumClosest.java
/*
 * Filename: 3SumClosest.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Wed 11 Mar 2015 04:45:05 PM EDT
 */

public class Solution {
	public int threeSumClosest(int[] num, int target) {
		Arrays.sort(num);

		int ret = (target > 0) ? Integer.MAX_VALUE : Integer.MIN_VALUE + 1;

		int n = num.length;
		for (int i = 0; i < n-2; i++) {
			if (i > 0 && num[i] == num[i-1]) {
				continue;
			}

			int a = num[i];
			int start = i+1;
			int end = n-1;

			while(start < end) {
				if (start != i+1 && num[start] == num[start-1]) {
					start++;
					continue;
				}

				if (end != n-1 && num[end] == num[end+1]) {
					end--;
					continue;
				}

				int b = num[start];
				int c = num[end];

				int sum = a + b + c;
				if (Math.abs(sum - target) < Math.abs(ret - target)) {
					ret = sum;
				}
				
				if (sum > target) {
					end--;
				} else if (sum < target) {
					start++;
				} else {
					return target;
				}
			}
		}

		return ret;
	}
}

File: cpp/3sum_closest.cc
/*
 * Filename: 3sum_closest.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 24 Jul 2015 01:32:24 PM EDT
 */

#include <algorithm>
#include <climits>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		int threeSumClosest(vector<int>& nums, int target) {
			sort(nums.begin(), nums.end());
			int diff = INT_MAX;
			int res;

			for (vector<int>::size_type i = 0; i < nums.size(); i++) {
				if (i && nums[i] == nums[i-1]) {
					continue;
				}

				int j = i+1;
				int k = nums.size() - 1;

				while (j < k) {
					if (j-i-1 && nums[j] == nums[j-1]) {
						j++;
						continue;
					}

					if (k < nums.size() - 1 && nums[k] == nums[k+1]) {
						k--;
						continue;
					}

					int sum = nums[i] + nums[j] + nums[k];

					if (abs(sum - target) < diff) {
						diff = abs(sum - target);
						res = sum;
					}

					if (sum == target) {
						return target;
					} else if (sum > target) {
						k--;
					} else {
						j++;
					}
				}
			}

			return res;
		}
};
